File: skills/code-architect/references/code-examples/ruby/state.rb
# frozen_string_literal: true

# State — let an object alter behavior when its state changes (state objects).

# Context delegates behavior to its current state object.
class Door
  attr_accessor :state

  def initialize = @state = ClosedState.new
  def push = @state.push(self)
  def pull = @state.pull(self)
  def label = @state.name
end

# Each state encapsulates behavior and the transitions it allows.
class ClosedState
  def name = "closed"
  def push(door) = door.state = OpenState.new
  def pull(_door) = nil # already closed
end

class OpenState
  def name = "open"
  def push(_door) = nil # already open
  def pull(door) = door.state = ClosedState.new
end

if __FILE__ == $PROGRAM_NAME
  door = Door.new
  puts door.label   # closed
  door.push
  puts door.label   # open
  door.push         # no-op, still open
  puts door.label
  door.pull
  puts door.label   # closed
end

File: skills/code-architect/references/code-examples/java/State.java
// State — let an object alter behavior when its state changes (state objects).

interface ConnectionState {
    ConnectionState open(Connection ctx);
    ConnectionState close(Connection ctx);
    String label();
}

class ClosedState implements ConnectionState {
    public ConnectionState open(Connection ctx) {
        System.out.println("opening connection");
        return new OpenState();
    }
    public ConnectionState close(Connection ctx) {
        System.out.println("already closed");
        return this;
    }
    public String label() { return "CLOSED"; }
}

class OpenState implements ConnectionState {
    public ConnectionState open(Connection ctx) {
        System.out.println("already open");
        return this;
    }
    public ConnectionState close(Connection ctx) {
        System.out.println("closing connection");
        return new ClosedState();
    }
    public String label() { return "OPEN"; }
}

class Connection {
    private ConnectionState state = new ClosedState();
    void open() { state = state.open(this); }
    void close() { state = state.close(this); }
    String status() { return state.label(); }
}

public class State {
    public static void main(String[] args) {
        Connection conn = new Connection();
        System.out.println("status: " + conn.status());
        conn.open();
        conn.open();
        System.out.println("status: " + conn.status());
        conn.close();
        System.out.println("status: " + conn.status());
    }
}
